text/0000-parent-trait-use.md

```diff
@@ -0,0 +1,176 @@
+- Feature Name: parent_trait_use
+- Start Date: 2018-01-18
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Allow implementing functions, associated constants and associated types of parent traits inside child traits.
+
+# Motivation
+[motivation]: #motivation
+
+This may have two kinds of effects: Better ergonomics for trait implementors and being able to change trait hierarchy by splitting types as a crate maintainer without breaking anything.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Assume there is a trait `A` which has parent traits `B` and `C`:
+```
+trait B {
+    fn b();
+}
+trait C {
+    fn c();
+}
+trait A: B+C {
+    fn a();
+}
+```
+Now you need to implement every of these traits for your new type `X` manually. But you only needs trait `A` and don't really care about `B` and `C`, maybe they are defined in some different crates, which you would need to depend on directly.
+So it's possible to implement every function directly inside the child trait like this:
+```
+struct X;
+impl A for X {
+    fn a() {…}
+    fn b() {…}
+    fn c() {…}
+}
+```
+But in order to allow this, trait `A` needs to be defined in a different way, like this:
+```
+trait A: use B+use C {
```

Comment:
It'd be cool if there was a way to specify that you want this behavior at the trait implementation, rather than the trait definition. This would allow you write these sorts of impls with existing traits.

Reply:
I don't think, it'd be useful at definition time.
What you think, would be something like that, I assume:
```
struct Number;
impl Add+Sub for Number {
    fn add(…) {…}
    fn sub(…) {…}
}
```
This may simplify implementation, but doesn't help with compatibility.

Maybe in combination with trait aliasses this would be useful.
When you want to split your trait `A` into `B` and `C`, like in the example from the RFC, you could write this:
```
trait B{…}
trait C{…}
trait A_old{…}
trait A = B+C+A_old;
```
This would be useful for backward compatibility too, but you would have to split it just into new traits without the ability to leave something inside the original trait. Also you probably couldn't create dynamic versions of this type then, anymore, which then would break it.

Reply:
My immediate reaction was that this seemed like a natural and neat extension which would save me a lot of boilerplate. However, I also think that boilerplate would be significantly reduced by existing (unimplemented) features such as implied bounds.

Reply:
I think implied bounds was the feature, that you don't have to write where clauses in every implementation, if you write it in the type definition itself, wasn't it?

Reply:
@porky11 Yes.